abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1996 @@
+[`proxy_get_header_map_pairs`] and/or [`proxy_set_header_map_pairs`]
+with `map_id` set to `HTTP_REQUEST_TRAILERS`.
+
+Individual HTTP request trailers can be retrieved and/or modified using
+[`proxy_get_header_map_value`], [`proxy_replace_header_map_value`],
+[`proxy_add_header_map_value`] and/or [`proxy_remove_header_map_value`]
+with `map_id` set to `HTTP_REQUEST_TRAILERS`.
+
+Paused request can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`HTTP_REQUEST`.
+
+Additionally, as long as HTTP response headers were not send downstream,
+a HTTP response can be sent using [`proxy_send_local_response`].
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `HTTP_REQUEST_TRAILERS` fields downstream.
+- `PAUSE` to pause processing.
+
+
+#### `proxy_on_response_headers`
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+  - `i32 (size_t) num_headers`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called when HTTP response headers are received from upstream.
+
+All `num_headers` headers can be retrieved and/or modified using
+[`proxy_get_header_map_pairs`] and/or [`proxy_set_header_map_pairs`]
+with `map_id` set to `HTTP_RESPONSE_HEADERS`.
+
+Individual headers can be retrieved and/or modified using
+[`proxy_get_header_map_value`], [`proxy_replace_header_map_value`],
+[`proxy_add_header_map_value`] and/or [`proxy_remove_header_map_value`]
+with `map_id` set to `HTTP_RESPONSE_HEADERS`.
+
+Paused request can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`HTTP_RESPONSE`.
+
+Additionally, instead of forwarding the response from upstream,
+a new HTTP response can be sent using [`proxy_send_local_response`].
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `HTTP_RESPONSE_HEADERS` fields downstream.
+- `PAUSE` to pause processing.
+
+
+#### `proxy_on_response_body`
+
+* params:
+  - `i32 (uint32_t) stream_context_id`
+  - `i32 (size_t) body_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (`[`proxy_action_t`]`) action`
+
+Called for each chunk of HTTP response body received from upstream,
+even when the processing is paused.
+
+Response body chunk (of `body_size`) can be retrieved and/or modified
+using [`proxy_get_buffer_bytes`] and/or [`proxy_set_buffer_bytes`]
+with `buffer_id` set to `HTTP_RESPONSE_BODY`.
+
+`body_size` represents total available size of the body that can be
+retrieved, and its value will increment if the processing is paused
+and the body is buffered by the host and not forwarded downstream.
+
+Paused HTTP response can be resumed using [`proxy_continue_stream`]
+or closed using [`proxy_close_stream`] with `stream_type` set to
+`HTTP_RESPONSE`.
+
+Plugin must return one of the following values:
+- `CONTINUE` to forward `HTTP_RESPONSE_BODY` buffer downstream.
+- `PAUSE` to pause processing.
+
```

Comment:
Can we clarify that we cannot use `proxy_send_local_response` at this point since the headers might have already arrived at the downstream?